### ISSUE TITLE
bug: s3 copy-object fails and does not generate a checksum

### ISSUE THREAD

**Issue by mmoales** [localstack/localstack#8172]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

In localstack, copying an s3 object in place and requesting a checksum causes an error

An error occurred (InvalidRequest) when calling the CopyObject operation: This copy request is illegal because it is trying to copy an object to itself without changing the object's metadata, storage class, website redirect location or encryption attributes.

### Expected Behavior

AWS s3 allows you to copy the object in place and request a checksum.

aws s3api put-object --bucket my-test-bucket --key go.mod --body .\go.mod
{
    "ETag": "\"f65b3f9b21e0c27be3c60df9c1cd87db\"",
    "ServerSideEncryption": "AES256"
}

aws s3api copy-object --copy-source my-test-bucket/go.mod --bucket my-test-bucket --key go.mod --storage-class STANDARD --checksum-algorithm SHA256
{
    "ServerSideEncryption": "AES256",
    "CopyObjectResult": {
        "ETag": "\"f65b3f9b21e0c27be3c60df9c1cd87db\"",
        "LastModified": "2023-04-19T17:09:16+00:00",
        "ChecksumSHA256": "nAPE7pxlr5L1D67nDJa1rwvmU3fKD2SuaUzeyhSR9Vc="
    }
}

### How are you starting LocalStack?

With a docker-compose file

### Steps To Reproduce

docker compose up -d

awslocal s3api create-bucket --bucket my-test-bucket

awslocal s3api put-object --bucket my-test-bucket --key go.mod --body .\go.mod

awslocal s3api copy-object --copy-source my-test-bucket/go.mod --bucket my-test-bucket --key go.mod --storage-class STANDARD --checksum-algorithm SHA256

### Environment

```markdown
- OS: Windows 11
- LocalStack: pro latest
```

**Comment by bentsku** [comment]
Hi @mmoales and thanks for your report!

After checking with AWS, it seems the issue is that we don't accept copy in place if the storage class is the same as the one currently set for the object, and requesting the checksum does not change anything on AWS side.
I'll check if the other fields keeping the same value raise the same error.

I'll work on a fix in the next days!

Simply because of curiosity, might I ask what is the use case of this?

**Comment by bentsku** [comment]
Hi @mmoales and thanks for the explanation!
So there's a different issue with returning the checksum for copy operation, I'll work on this one, this should be quicker to fix if that can unblock you.

I started working on a fix for the "in-place" issue but it will take a bit more time as we need to change some assumptions.
Thanks again!

**Comment by mmoales** [comment]
Hi @bentsku,

Thanks for the quick response! Unfortunately, we really don't want to change the storage class when we do the copy. If simply adding metadata would prevent the error, we could probably get by with doing that. However, copy-in-place fails even if I supply metadata:

awslocal s3api copy-object --copy-source cds-storage-local-data-bucket/go.mod --bucket cds-storage-local-data-bucket --key go.mod --checksum-algorithm SHA256 --metadata keyname=string

An error occurred (InvalidRequest) when calling the CopyObject operation: This copy request is illegal because it is trying to copy an object to itself without changing the object's metadata, storage class, website redirect location or encryption attributes.

**Comment by bentsku** [comment]
Great, I can see our CopyObject check is really not working well. I think there might be workarounds for now, depending on your use case, but they're all bad until this is fixed: you can try setting `--metadata-directive=REPLACE` and S3 will accept your request, even without the storage class or metadata. This is due to an error in the logic on how S3 should handle these. However, it would effectively erase the metadata of yours objects...

Here's the check that will need refactoring:
```python
if (
    src_key.name == dest_key_name
    and src_key.bucket_name == dest_bucket_name
    and storage == src_key.storage_class
    and acl == src_key.acl
    and encryption == src_key.encryption
    and kms_key_id == src_key.kms_key_id
    and bucket_key_enabled == (src_key.bucket_key_enabled or False)
    and mdirective != "REPLACE"
):
    raise CopyObjectMustChangeSomething
```

I'll keep you posted on the progress!

**Comment by mmoales** [comment]
Hi @bentsku, thanks again for addressing this. Should I file a separate issue for the checksum not being generated on a copy operation?

**Comment by bentsku** [comment]
It's okay for the 2nd issue, I'm already working on it, but thanks for asking! 👍

**Comment by bentsku** [comment]
Hello @mmoales, the fix should now be merged and available, the linked PR was only the tests validating the behaviour. You can give it a try by pulling the `latest` docker image. Thanks a lot for your report, it led to a lot of good changes along the way!

**Comment by mmoales** [comment]
@bentsku Excellent! I have tried out the latest changes and see that it's working. Thanks for you help!
